rust/src/storage/file/file_backend/multi_page_cached_file.rs

```diff
@@ -0,0 +1,284 @@
+        let mut file = MockFileBackend::new();
+        file.expect_write_all_at()
+            .withf(|buf, offset| {
+                buf == [0; Page::SIZE] && (*offset == 0 || *offset == Page::SIZE as u64)
+            })
+            .times(1)
+            .returning(|_, _| Ok(()));
+        file.expect_read_exact_at()
+            .withf(|_, offset| *offset == 2 * Page::SIZE as u64)
+            .times(1)
+            .returning(|buf, _| {
+                buf.fill(1);
+                Ok(())
+            });
+
+        let file = MultiPageCachedFile::<_, _, true> {
+            file,
+            file_len: AtomicU64::new(3 * Page::SIZE as u64),
+            pages: Pages::new([(Page::zeroed(), 0), (Page::zeroed(), Page::SIZE as u64)]),
+        };
+        // Mark first page as dirty by writing to it.
+        file.write_all_at(&[0], 0).unwrap();
+
+        // Access data outside of the cached pages, which should trigger a write of the old page and
+        // a read of the new page.
+        let mut read_data = vec![0u8; Page::SIZE];
+        file.read_exact_at(&mut read_data, 2 * Page::SIZE as u64)
+            .unwrap();
+        assert_eq!(read_data, vec![1u8; Page::SIZE]);
+    }
+
+    #[test]
+    fn accessing_cached_page_blocks_until_this_page_becomes_available() {
+        let file = MockFileBackend::new();
+        let file = MultiPageCachedFile::<_, _, true> {
+            file,
+            file_len: AtomicU64::new(2 * Page::SIZE as u64),
+            pages: Pages::new([(Page::zeroed(), 0), (Page::zeroed(), 1)]),
+        };
+        let file = Arc::new(file);
+
+        let _locked_page = file.change_page(0).unwrap();
+
+        // Try to access the same page from another thread. This should block until the first lock
+        // is dropped.
+        let handle = std::thread::spawn({
+            let file = Arc::clone(&file);
+            move || {
+                let _locked_page = file.change_page(0).unwrap();
+            }
+        });
+
+        std::thread::sleep(std::time::Duration::from_millis(100));
+        assert!(!handle.is_finished());
+
+        drop(_locked_page);
+
+        std::thread::sleep(std::time::Duration::from_millis(100));
+        assert!(handle.is_finished());
+    }
+
+    #[test]
+    fn accessing_non_cached_page_blocks_page_becomes_available() {
+        let file = MockFileBackend::new();
+        let file = MultiPageCachedFile::<_, _, true> {
+            file,
+            file_len: AtomicU64::new(3 * Page::SIZE as u64),
+            pages: Pages::new([(Page::zeroed(), 0), (Page::zeroed(), 1)]),
+        };
+        let file = Arc::new(file);
+
+        // Lock the first cached page.
+        let locked_page1 = file.change_page(0).unwrap();
+
+        // Locking the second cached page should succeed immediately.
+        assert!(file.change_page(Page::SIZE as u64).is_ok());
+
+        // Locking the first page should block until the guard to it is dropped.
```

Comment:
```suggestion
        // Locking the third page should block until the guard to page 1 is dropped,
        // since it occupies the same cache slot.
```